text/Issue 8 - Plug-in based architecture.md

```diff
@@ -0,0 +1,265 @@
+a set of `ChannelKind` and on (de)serialization format.
+
+Having such enums saves time to adapter developers, who do not need to write
+their own (de)serializers.
+
+Having such enums means that we can dispatch the same `Value` to different
+adapters that offer the same feature, without every single adapter having to
+reparse the same `Value`.
+
+Having such enums is good for optimization of the foxbox.
+
+Having such enums makes it possible to make big changes/fixes to the entire architecture
+without touching the code of adapters. For instance, ongoing work on
+https://github.com/fxbox/taxonomy/issues/37 demonstrates that we can entirely
+change the (de)serialization format, or even provide several formats that can
+be picked by users at runtime, without the need to patch every single `Adapter`.
+
+## Cons
+
+Decentralized is more webby – for better or worse, this is what most developers
+expect.
+
+Experience shows that centralizing everything hurts extensibility and adoption.
+
+We will eventually end up with thousands of variants for these enums, which
+is bad for readability, documentation, testing, etc.
+
+This adds a non-negligible maintenance burden to the Taxonomy repo.
+
+## Bottom line
+
+We need to find a solution that:
+
+- lets us standardize channel kinds whenever possible;
+- abstracts away (de)serialization whenever possible;
+- lets us optimize;
+- does not require centralization.
+
+# Plug-in based architecture
+
+The general idea behind this proposal is to keep types that will help discoverability
+and (de)serialization, but maintain them as plug-ins instead of centralized enums.
+The Foxbox can come with a number of bundled plug-ins, which represent
+standardized features, Adapters can come with other plug-ins, and plug-ins
+that prove successful can progressively be moved from the code of Adapters
+to the standard bundle.
+
+
+## Values, serialization, deserialization
+
+We replace `Value` and `Type` by a new, extensible, trait `Message`. The (only)
+role of a `Message` is to standardize (de)serialization.
+
+A crate `standardized` provides a library of standard `Message`s to cover most uses:
+- boolean;
+- string;
+- various number types;
+- dates;
+- time;
+- temperature;
+- image;
+- `RawJSON`;
+- `RawBinary`;
+- ...
+
+Use of a `Message` is decided per-channel, by the Adapter developer.
+
+When implementing a channel, adapter developers are encouraged to pick one
+of the full-featured `Message`s (e.g. temperature or time), or can also provide
+new implementations of `Message`, if they prefer. Additional instances of `Message`
+can easily be published by third-party developers on third-party crates.
+
+While this is less optimal, Adapter developers can fallback to `RawJSON` or `RawBinary`
+if they prefer implementing their own mechanism internally, although this may
+come at some cost performance cost (typically, they may need to reparse the
+same value several times from JSON or Binary).
+
+
```

Comment:
I like this first section, in that it encapsulates both Type and Value into one object that can be passed around somewhat opaquely.